### ISSUE TITLE
Proposal: (Partial) Language-specific Pass Replacement

### ISSUE THREAD

**Issue by oxisto** [Fraunhofer-AISEC/cpg#1161]
In some languages, we need (or we may) fine-tune to EOG or create EOG edges that do not exist in other languages. Some examples are the handling of default arguments (which might be very different in languages) or language concepts that are sort of unique, such as `defer` in Go (which executes the supplied call expression after exiting the function body).

I would envision something like the following. First, I am creating a language-specific EOG class (which derives from the regular EOG pass).

```kotlin
class GoEOG : EvaluationOrderGraphPass() {
  override fun handleFunctionDeclaration(node: FunctionDeclaration) {
    // handle defer stuff
    /* ... */

    // pass to regular EOG
    super.handleFunctionDeclaration(node)
  }
}
```

Inside a language frontend, I can then use an annotation similar to the existing `RegisterExtraPass`, for example `ReplacePass` or maybe `RegisterReplacementPass`.

```kotlin
@ReplacePass(EvaluationOrderGraphPass::class, for = GoLanguage::class, with = GoEOG::class)
class GoLanguageFrontend : LanguageFrontend() {}
```

The tricky part would be that we would need to execute the pass only for software components / or translation units (?) that are of `GoLanguage`, for all others, we would need to execute the regular EOG. Not sure yet how to do that exactly.

Any opinions on that?

cc @Fraunhofer-AISEC/sig-software-security

**Comment by KuechA** [comment]
Currently, we model some of these differences with `LanguageTrait`s. But since this might end up in a very fine-grained configuration of very specific behaviors, it could indeed make sense to offer such a functionality. I agree that the implementation of such a feature might be tricky and probably the "main pass" (e.g. the `EvaluationOrderGraphPass` in your example) will have to pick the correct subclass based on which refined passes have been registered and the language of the current node.

**Comment by oxisto** [comment]
> Currently, we model some of these differences with `LanguageTrait`s. But since this might end up in a very fine-grained configuration of very specific behaviors, it could indeed make sense to offer such a functionality. I agree that the implementation of such a feature might be tricky and probably the "main pass" (e.g. the `EvaluationOrderGraphPass` in your example) will have to pick the correct subclass based on which refined passes have been registered and the language of the current node.

I would probably do that at the level of a `Component`.  Currently every pass is executed across all components, this is strictly speaking not such a good idea as components should model some kind of logical boundary between different parts, e.g. an application and a library; or a client and server application. Passes should probably executed *once per* component, and then we can execute the passes based on the language of the component.

**Comment by KuechA** [comment]
Ok, so this actually makes two issues:
1) The language-specific passes and
2) running a pass for each `Component` instead of the `TranslationResult`

Concerning 2) I guess we probably want to change the `accept(t: TranslationResult)` to `accept(c: Component)` and split the TR into its components and pass these to this method (which could also help parallelizing the passes)

**Comment by oxisto** [comment]
Just to reflect on the discussion today:

- We will introduce a way to specify the "granularity" of a pass, i.e., translation result, software component or translation unit. Probably through the use of dedicated interface/abstract classes like `ResultPass`
- We will make passes replaceable at all the different levels - although it might only work for the granularity level which actually has a specified `language` - so probably only at the translation unit for now.

**Comment by KuechA** [comment]
Closed by  #1169
